### DISCUSSION TITLE
imagemagick mp4 to gif command question

### DISCUSSION THREAD

**Discussion by 0lm** [ImageMagick/ImageMagick#5170] [Help]
hi!

in the very past i used the following command to turn a few webp files into a gif files:

`convert -delay 10 -dispose none *.webp -coalesce -loop 0 -layers optimize converted%d.gif`

now i have a few mp4 files (each is 5-15sec long and 1080p) and I used exactly the same command, just replaced *.webp by *.mp4

well...now imagemagick is working since 1h and still no results, and it conatantly uses 80% - 99%  of my RAM all the time. and my available drive space also drops (like.. when i had 900 GB space, I have 860 GB now and it drops more and more by time).

well i imagine i did something stupid. but i dont fully understand why. why is the command i used to convert mp4 to gif not good? why does it take so long and why does it eat up so much drive space and RAM?

should i cancle the conversion? if i do, where do I usually find the temp files which were already created? (30 GB+ big, i would like to delete them).

my initial idea was: to convert the mp4 clips 1:1 to gif (basically lossless, same quality with same resolution). which command sequence is recommended to get losless gifs? - if this is too much of a work for CPU's / computers, then: i dont mind to make resolution smaller. but i still would like same good quality, just in smaller resolution and in the best case without skipping any frames. which command sequence would be the best for this?

**Comment by fmw42** [comment]
Your MP4 file probably has too many frames to read all frame into memory.  Either you do not have enough RAM for that or you are thrashing to disk. You may need to modify your policy.xml file for large amount of input images.

**Comment by snibgo** [comment]
What version of IM, on what platform?

Perhaps your command is processing hundreds of large images, and they exceed the capacity of your memory, so IM is using disk instead, which is much slower.

Temporary files by default are in %TEMP% in Windows, or /tmp on *nix systems.

I suggest you convert just one image, by using [0] after the input mp4 filename. How large is that image, in pixels? How long does it take?

You might convert just one image at a time, or perhaps 10 images at a time, or whatever doesn't exceed your memory.

**Comment by 0lm** [comment]
Thank you for your replies, both of you.

Yes I had the feeling IM tries to process too many frames at once. The videos (even if they are short, only 5-15sec each) are rather high quality with probably alot of frames at 1080p.

My sytsem is windows 10 64bit, 16 GB DDR4 ram, i was using the recommended newest IM 7 for windows from IM's main page (literally the newest. updated it today hours ago)

The idea of making IM process only a few images at a time sounds interesting. So the issue is, IM processes limitless numbers of frames right now?

and adding 0... would it look like this?
`
convert -delay 10 -dispose none *[0].mp4 -coalesce -loop 0 -layers optimize converted%d.gif`

i dont really have a file name, since its multiple files, thats why I made this simple bulk converting command. i basically open cmd in the folder where the mp4 files are I want to convert, and execute this command. so would i need to add the [0] at that position then? will it still work in batch? and adding [10] instead will process 10 images at once only, and then the next 10 after they are done, etc?

**Comment by snibgo** [comment]
So you are reading *all* of the frames from *all* of your video files into memory at the same time. Oops.

Try it with just one video file. If that works, put that in a shell "for" loop, to repeat the IM command for each file in turn.

But if it doesn't work, try just one image from one file. For example:
```
yourfilename.mp4[0]
```
... to read just the first frame. Or...
```
yourfilename.mp4[0-9]
```
to read just the first 10 frames.

**Comment by 0lm** [comment]
i see, so thats what happening in the background lol. i actually assumed IM would do the files 1 by 1 automatically.

the issue is, i dont know how many frames each video has. so when i restrict it by [0] or [0-9], does it mean it will only turn a single frame (which would by  a standing still image) into gif? and in the 2nd example, only the first 10 frames into a gif?

**Reply by fmw42** [reply]
Process each mp4 separately into a multi-frame gif. All input frames will be stored in the output gif. Then combine all the gifs into one gif.  If necessary, read each mp4 in groups of frames using the [a-b] syntax. You may want to write a script loop over each mp4 and a loop to read only so many frames at a time.

**Comment by snibgo** [comment]
For the number of frames in a MP4 file, I use ffprobe. (Part of the ffmpeg set.)
```
ffprobe -show_streams sample.mp4
```
This gives text data, including:
```
width=1920
height=1080
r_frame_rate=60000/1001
duration=7.640967
```
This movie has 60000/1001 = 59.94... frames per seconds, and is 7.64... seconds long.

So the number of frames is 7.640967*60000/1001 = 458. They have about 2m pixels per frame, so about 1 billion pixels.

My laptop has 12 GB memory, which can easily cope with 20m pixels, and probably not 1b pixels. So I would use a shell loop (Windows "for /L") that processes the first 10 frames, then the next 10, etc.

**Comment by snibgo** [comment]
Incidentally, if all you want to do is create an animated GIF from a MP4, I suggest that ffmpeg is a far more suitable tool than ImageMagick.
